File: pb/src/main/java/org/sqlbuilder/pb/Exporter.java
package org.sqlbuilder.pb;

import org.sqlbuilder.common.Names;
import org.sqlbuilder.common.ProvidesIdTo;
import org.sqlbuilder.pb.objects.Role;
import org.sqlbuilder.pb.objects.RoleSet;
import org.sqlbuilder.pb.objects.Theta;
import org.sqlbuilder.pb.objects.Word;
import org.sqlbuilder2.ser.Pair;
import org.sqlbuilder2.ser.Serialize;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.util.AbstractMap.SimpleEntry;
import java.util.Comparator;
import java.util.Map;
import java.util.Properties;
import java.util.TreeMap;
import java.util.function.Function;

import static java.util.stream.Collectors.*;

public class Exporter
{
	//private static final Comparator<Pair<String, String>> NONSERIALIZABLE_COMPARATOR = Comparator.comparing(Pair<String, String>::getFirst).thenComparing(Pair::getSecond);
	//private static final Comparator<Pair<String, String>> COMPARATOR = (Comparator<Pair<String, String>> & Serializable) (p1, p2) -> NONSERIALIZABLE_COMPARATOR.compare(p1, p2);

	private static final Comparator<Pair<String, String>> COMPARATOR = Comparator.comparing(Pair<String, String>::getFirst).thenComparing(Pair::getSecond);

	protected final Names names;

	protected final File outDir;

	public Exporter(final Properties conf)
	{
		this.names = new Names("pb");
		this.outDir = new File(conf.getProperty("pb_outdir_ser", "sers"));
		if (!this.outDir.exists())
		{
			//noinspection ResultOfMethodCallIgnored
			this.outDir.mkdirs();
		}
	}

	public void run() throws IOException
	{
		System.out.printf("%s %d%n", "thetas", Theta.COLLECTOR.size());
		System.out.printf("%s %d%n", "roles", Role.COLLECTOR.size());
		System.out.printf("%s %d%n", "classes", RoleSet.COLLECTOR.size());
		System.out.printf("%s %d%n", "words", Word.COLLECTOR.size());
		duplicateRoles();

		try ( //
		      @ProvidesIdTo(type = Theta.class) var ignored1 = Theta.COLLECTOR.open(); //
		      @ProvidesIdTo(type = Role.class) var ignored2 = Role.COLLECTOR.open(); //
		      @ProvidesIdTo(type = RoleSet.class) var ignored3 = RoleSet.COLLECTOR.open(); //
		      @ProvidesIdTo(type = Word.class) var ignored4 = Word.COLLECTOR.open() //
		)
		{
			serialize();
			export();
		}
	}

	public void serialize() throws IOException
	{
		serializeThetas();
		serializeRoleSets();
		serializeRolesBare();
		serializeRoles();
		serializeWords();
	}

	public void export() throws IOException
	{
		exportThetas();
		exportRoleSets();
		exportRolesBare();
		exportRoles();
		exportWords();
	}

	public void serializeThetas() throws IOException
	{
		var m = makeThetasMap();
		Serialize.serialize(m, new File(outDir, names.serFile("thetas.resolve", "_[theta]-[thetaid]")));
	}

	public void serializeRoleSets() throws IOException
	{
		var m = makeRoleSetsMap();
		Serialize.serialize(m, new File(outDir, names.serFile("rolesets.resolve", "_[roleset]-[rolesetid]")));
	}

	public void serializeRolesBare() throws IOException
	{
		var m = makeRolesMap();
		Serialize.serialize(m, new File(outDir, names.serFile("roles.resolve", "_[roleset,argn]-[roleid]")));
	}

	private void serializeRoles() throws IOException
	{
		var m = makeRolesFromArgNToFullMap();
		Serialize.serialize(m, new File(outDir, names.serFile("roles.resolve", "_[roleset,argn]-[roleid,rolesetid]")));
	}

	public void serializeWords() throws IOException
	{
		var m = makeWordMap();
		Serialize.serialize(m, new File(outDir, names.serFile("words.resolve", "_[word]-[pbwordid]")));
	}

	public void exportThetas() throws IOException
	{
		var m = makeThetasMap();
		export(m, new File(outDir, names.mapFile("thetas.resolve", "_[theta]-[thetaid]")));
	}

	public void exportRoleSets() throws IOException
	{
		var m = makeRoleSetsMap();
		export(m, new File(outDir, names.mapFile("rolesets.resolve", "_[roleset]-[rolesetid]")));
	}

	public void exportRolesBare() throws IOException
	{
		var m = makeRolesTreeMap();
		export(m, new File(outDir, names.mapFile("roles.resolve", "_[roleset,argn]-[roleid]")));
	}

	public void exportRoles() throws IOException
	{
		var m = makeRolesFromArgNToFullTreeMap();
		export(m, new File(outDir, names.mapFile("roles.resolve", "_[roleset,argn]-[roleid,rolesetid]")));
	}

	public void exportWords() throws IOException
	{
		var m = makeWordMap();
		export(m, new File(outDir, names.mapFile("words.resolve", "_[word]-[pbwordid]")));
	}

	public static <K, V> void export(Map<K, V> m, File file) throws IOException
	{
		try (PrintStream ps = new PrintStream(new FileOutputStream(file), true, StandardCharsets.UTF_8))
		{
			export(ps, m);
		}
	}

	public static <K, V> void export(PrintStream ps, Map<K, V> m)
	{
		m.forEach((strs, nids) -> ps.printf("%s -> %s%n", strs, nids));
	}

	void duplicateRoles()
	{
		Role.COLLECTOR.keySet().stream() //
				.map(r -> new Pair<>(r.getRoleSet().getName(), r.getArgn())) //
				.collect(groupingBy(Function.identity(), counting())) //
				.entrySet().stream() //
				.filter(e -> e.getValue() > 1).map(Map.Entry::getKey).forEach(p -> System.err.printf("%s is duplicated in %s%n", p.second, p.first));
	}

	// M A P S

	/**
	 * Make word to wordid map
	 *
	 * @return word to wordid
	 */
	public Map<String, Integer> makeWordMap()
	{
		return Word.COLLECTOR.entrySet().stream() //
				.map(e -> new SimpleEntry<>(e.getKey().getWord(), e.getValue())) //
				.collect(toMap(SimpleEntry::getKey, SimpleEntry::getValue, (x, r) -> x, TreeMap::new));
	}

	public Map<String, Integer> makeRoleSetsMap()
	{
		return RoleSet.COLLECTOR.entrySet().stream() //
				.map(e -> new SimpleEntry<>(e.getKey().getName(), e.getValue())) //
				.collect(toMap(SimpleEntry::getKey, SimpleEntry::getValue, (x, r) -> x, TreeMap::new));
	}

	public Map<String, Integer> makeThetasMap()
	{
		return Theta.COLLECTOR.entrySet().stream() //
				.map(e -> new SimpleEntry<>(e.getKey().getTheta(), e.getValue())) //
				.collect(toMap(SimpleEntry::getKey, SimpleEntry::getValue, (x, r) -> x, TreeMap::new));
	}

	public Map<Pair<String, String>, Integer> makeRolesMap()
	{
		return Role.COLLECTOR.entrySet().stream() //
				.map(e -> {
					var r = e.getKey(); //
					var rs = r.getRoleSet();
					return new Pair<>( //
							new Pair<>(rs.getName(), r.getArgn()), //
							e.getValue());
				}) //
				.collect(toMap(Pair::getFirst, Pair::getSecond, (x, r) -> {
					System.err.println("duplicate roleid for roleset: " + x + " / " + r);
					return x;
				}));
	}

	public Map<Pair<String, String>, Integer> makeRolesTreeMap()
	{
		return Role.COLLECTOR.entrySet().stream() //
				.map(e -> {
					var r = e.getKey(); //
					var rs = r.getRoleSet();
					return new Pair<>( //
							new Pair<>(rs.getName(), r.getArgn()), //
							e.getValue());
				}) //
				.collect(toMap(Pair::getFirst, Pair::getSecond, (x, r) -> {
					// System.err.println("duplicate roleid for roleset: " + x + " / " + r);
					return x;
				}, () -> new TreeMap<>(COMPARATOR)));
	}

	/**
	 * Detailed role maps
	 *
	 * @return (rolesetname, argn) -> (roleid, rolesetid)
	 */
	public Map<Pair<String, String>, Pair<Integer, Integer>> makeRolesFromArgNToFullMap()
	{
		return Role.COLLECTOR.entrySet().stream() //
				.map(e -> {
					var r = e.getKey(); //
					var rs = r.getRoleSet();
					return new Pair<>( //
							new Pair<>(rs.getName(), r.getArgn()), //
							new Pair<>(e.getValue(), rs.getIntId()));
				}) //
				.collect(toMap(Pair::getFirst, Pair::getSecond, (x, r) -> {
					// System.err.println("duplicate roleid for roleset: " + x + " / " + r);
					return x;
				}));
	}

	/**
	 * Detailed role maps
	 *
	 * @return (rolesetname, argn) -> (roleid, rolesetid)
	 */
	public Map<Pair<String, String>, Pair<Integer, Integer>> makeRolesFromArgNToFullTreeMap()
	{
		return Role.COLLECTOR.entrySet().stream() //
				.map(e -> {
					var r = e.getKey(); //
					var rs = r.getRoleSet();
					return new Pair<>( //
							new Pair<>(rs.getName(), r.getArgn()), //
							new Pair<>(e.getValue(), rs.getIntId()));
				}) //
				.collect(toMap(Pair::getFirst, Pair::getSecond, (x, r) -> {
					// System.err.println("duplicate roleid for roleset: " + x + " / " + r);
					return x;
				}, () -> new TreeMap<>(COMPARATOR)));
	}
}

File: pb/src/main/java/org/sqlbuilder/pb/Exporter.kt
package org.sqlbuilder.pb

import org.sqlbuilder.common.Names
import org.sqlbuilder.pb.foreign.AliasFnFeLinks
import org.sqlbuilder.pb.foreign.AliasVnRoleLinks
import org.sqlbuilder.pb.objects.Role
import org.sqlbuilder.pb.objects.RoleSet
import org.sqlbuilder.pb.objects.Word
import org.sqlbuilder2.ser.Serialize
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.PrintStream
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.Throws

open class Exporter(conf: Properties) {

    protected val names: Names = Names("pb")

    protected val outDir: File = File(conf.getProperty("pb_outdir_ser", "sers"))

    init {
        if (!outDir.exists()) {
            outDir.mkdirs()
        }
    }

    @Throws(IOException::class)
    fun run() {
        System.out.printf("%s %d%n", "vn links", AliasFnFeLinks.COLLECTOR.size)
        System.out.printf("%s %d%n", "fn links", AliasFnFeLinks.COLLECTOR.size)
        System.out.printf("%s %d%n", "roles", Role.COLLECTOR.size)
        System.out.printf("%s %d%n", "classes", RoleSet.COLLECTOR.size)
        System.out.printf("%s %d%n", "words", Word.COLLECTOR.size)
        duplicateRoles()

        AliasFnFeLinks.COLLECTOR.open().use {
            AliasVnRoleLinks.COLLECTOR.open().use {
                Role.COLLECTOR.open().use {
                    RoleSet.COLLECTOR.open().use {
                        Word.COLLECTOR.open().use {
                            serialize()
                            export()
                        }
                    }
                }
            }
        }
    }

    @Throws(IOException::class)
    fun serialize() {
        serializeVnRoles()
        serializeFnFes()
        serializeRoleSets()
        serializeRolesBare()
        serializeRoles()
        serializeWords()
    }

    @Throws(IOException::class)
    fun export() {
        exportVnRoles()
        exportFnFes()
        exportRoleSets()
        exportRolesBare()
        exportRoles()
        exportWords()
    }

    @Throws(IOException::class)
    fun serializeWords() {
        val m = makeWordMap()
        Serialize.serialize(m, File(outDir, names.serFile("words", ".resolve_[word]-[pbwordid]")))
    }

    @Throws(IOException::class)
    fun exportWords() {
        val m = makeWordMap()
        export(m, File(outDir, names.mapFile("words.resolve", "_[word]-[pbwordid]")))
    }

    @Throws(IOException::class)
    fun serializeRoleSets() {
        val m = makeRoleSetsMap()
        Serialize.serialize(m, File(outDir, names.serFile("rolesets", ".resolve_[roleset]-[rolesetid]")))
    }

    @Throws(IOException::class)
    fun exportRoleSets() {
        val m = makeRoleSetsMap()
        export(m, File(outDir, names.mapFile("rolesets.resolve", "_[roleset]-[rolesetid]")))
    }

    @Throws(IOException::class)
    private fun serializeRoles() {
        val m = makeRolesFromArgTypeToFullMap()
        Serialize.serialize(m, File(outDir, names.serFile("roles", ".resolve_[roleset,argtype]-[roleid,rolesetid]")))
    }

    @Throws(IOException::class)
    fun exportRoles() {
        val m = makeRolesFromArgTypeToFullTreeMap()
        export(m, File(outDir, names.mapFile("roles.resolve", "_[roleset,argtype]-[roleid,rolesetid]")))
    }

    @Throws(IOException::class)
    fun serializeRolesBare() {
        val m = makeRolesMap()
        Serialize.serialize(m, File(outDir, names.serFile("roles", ".resolve_[roleset,argtype]-[roleid]")))
    }

    @Throws(IOException::class)
    fun exportRolesBare() {
        val m = makeRolesTreeMap()
        export(m, File(outDir, names.mapFile("roles.resolve", "_[roleset,argtype]-[roleid]")))
    }

    @Throws(IOException::class)
    fun serializeVnRoles() {
        val m = makeVnRolesMap()
        Serialize.serialize(m, File(outDir, names.serFile("vnroles", ".resolve_[fe]-[feid]")))
    }

    @Throws(IOException::class)
    fun exportVnRoles() {
        val m = makeVnRolesMap()
        export(m, File(outDir, names.mapFile("vnroles.resolve", "_[vnrole]-[vnroleid]")))
    }

    @Throws(IOException::class)
    fun serializeFnFes() {
        val m = makeFnFesMap()
        Serialize.serialize(m, File(outDir, names.serFile("fnfes", ".resolve_[fe]-[feid]")))
    }

    @Throws(IOException::class)
    fun exportFnFes() {
        val m = makeFnFesMap()
        export(m, File(outDir, names.mapFile("fnfes.resolve", "_[fe]-[feid]")))
    }

    fun duplicateRoles() {
        Role.COLLECTOR
            .groupBy { it }
            .forEach { (role, instances) ->
                if (instances.size > 1) {
                    println("Warning: Duplicate role: ${instances.size} instances for  role '$role' (rs=${role.roleSet} argt=${role.argType}) found.")
                }
            }
    }

    // M A P S

    /**
     * Make word to wordid map
     *
     * @return word to wordid
     */
    fun makeWordMap(): Map<String, Int> {
        return Word.COLLECTOR
            .asSequence()
            .map { it.word to Word.COLLECTOR.apply(it) }
            .toMap()
            .toSortedMap()
    }

    fun makeRoleSetsMap(): Map<String, Int> {
        return RoleSet.COLLECTOR
            .asSequence()
            .map { it.name to RoleSet.COLLECTOR.apply(it) }
            .toMap()
            .toSortedMap()
    }

    fun makeVnRolesMap(): Map<String, Int> {
        return AliasVnRoleLinks.COLLECTOR
            .asSequence()
            .map { it.names.toString() to AliasVnRoleLinks.COLLECTOR.apply(it) }
            .toMap()
            .toSortedMap()
    }

    fun makeFnFesMap(): Map<String, Int> {
        return AliasFnFeLinks.COLLECTOR
            .asSequence()
            .map { it.names.toString() to AliasFnFeLinks.COLLECTOR.apply(it) }
            .toMap()
            .toSortedMap()
    }

    fun makeRolesMap(): Map<Pair<String, String>, Int> {
        duplicateRoles()
        return Role.COLLECTOR
            .asSequence()
            .map {
                val rs = it.roleSet
                (rs.name to it.argType) to Role.COLLECTOR.apply(it)
            }
            .toMap()
    }

    fun makeRolesTreeMap(): Map<Pair<String, String>, Int> {
        return Role.COLLECTOR
            .asSequence()
            .map {
                val rs = it.roleSet
                (rs.name to it.argType) to Role.COLLECTOR.apply(it)
            }
            .toMap()
            .toSortedMap(COMPARATOR)
    }

    /**
     * Detailed role maps
     *
     * @return (rolesetname, argtype) -> (roleid, rolesetid)
     */
    fun makeRolesFromArgTypeToFullMap(): Map<Pair<String, String>, Pair<Int, Int>> {
        return Role.COLLECTOR
            .asSequence()
            .map {
                val rs = it.roleSet
                (rs.name to it.argType) to (Role.COLLECTOR.apply(it) to rs.intId)
            }
            .toMap()
    }

    /**
     * Detailed role maps
     *
     * @return (rolesetname, argtype) -> (roleid, rolesetid)
     */
    fun makeRolesFromArgTypeToFullTreeMap(): Map<Pair<String, String>, Pair<Int, Int>> {
        return Role.COLLECTOR
            .asSequence()
            .map {
                val rs = it.roleSet
                (rs.name to it.argType) to (Role.COLLECTOR.apply(it) to rs.intId)
            }
            .toMap()
            .toSortedMap(COMPARATOR)
    }

    companion object {

        private val COMPARATOR: Comparator<Pair<String, String>> =
            Comparator.comparing { p: Pair<String, String> -> p.first }
                .thenComparing<String> { it.second }

        @Throws(IOException::class)
        fun <K, V> export(m: Map<K, V>, file: File) {
            PrintStream(FileOutputStream(file), true, StandardCharsets.UTF_8).use { ps ->
                export<K, V>(ps, m)
            }
        }

        fun <K, V> export(ps: PrintStream, m: Map<K, V>) {
            m.forEach { ps.println("${it.key} -> ${it.value}") }
        }
    }
}